reps/2023-12-14-virtual-cluster.md

```diff
@@ -0,0 +1,452 @@
+  // of resources before starting the cluster.
+  // If not specified, the default value is 0.
+  map<string, double> flexible_resource_min
+
+  // If specified, limit the consumption of these resources to
+  // the specified values.
+  // If not specified, the default value is infinite.
+  map<string, double> flexible_resource_max
+
+  // == Fixed size resources ==
+  // Fixed sized virtual nodes to request, e.g. {"GPU": 4}.
+  // These virtual node resources are part of the min resources
+  // together with flexible_resource_min
+  // that will be atomically reserved when the
+  // virtual cluster is created.
+  repeated FixedSizeNodes fixed_size_nodes
+}
+
+message LabelIn {
+  repeated string values = 1;
+}
+
+message LabelNotIn {
+  repeated string values = 1;
+}
+
+message LabelExists {}
+
+message LabelDoesNotExist {}
+
+message LabelOperator {
+  oneof label_operator {
+    LabelIn label_in = 1;
+    LabelNotIn label_not_in = 2;
+    LabelExists label_exists = 3;
+    LabelDoesNotExist label_does_not_exist = 4;
+  }
+}
+
+message LabelMatchExpression {
+  string key = 1;
+  LabelOperator operator = 2;
+}
+
+message LabelMatchExpressions {
+  repeated LabelMatchExpression expressions = 1;
+}
+
+message FixedSizeNode {
+  map<string, double> resources
+
+  // Use label match expression to
+  // specify candidate parent nodes
+  // of this fixed size node.
+  LabelMatchExpressions parent_node_selector
+
+  // Additional labels that the
+  // virtual node has in addition to
+  // those inherited from the parent node.
+  map<string, string> labels
+}
+
+enum SchedulingPolicy {
+  // Same as the current placement group strategies.
+  PACK
+  SPREAD
+  STRICT_SPREAD
+}
+
+message FixedSizeNodes {
+  // This can be used to represent placement groups
+  // where each bundle is a FixedSizeNode and
+  // the scheduling policy is the placement group strategy.
+
+  repeated FixedSizeNode nodes
+  // One of PACK, SPREAD, or STRICT_SPREAD. These would be
+  // defined with respect to parent virtual nodes for nested
+  // clusters.
+  SchedulingPolicy scheduling_policy
+}
+```
+
+#### Job API
+
+Currently we have two ways to run a Ray job: ``ray.init()`` and ``ray job submit``. Both will take an optional parameter specifying the spec of the virtual cluster inside which the job will run. If unspecified, the default virtual cluster has zero min resources and infinite max resources meaning it can scale up to use the entire physical cluster resources.
+
+```python
+# Default virtual cluster
+# The job can use up to the entire physical cluster resources.
+ray.init()
+
+# The job can use at least 1 CPU and at most 8 CPUs.
+ray.init(virtual_cluster=VirtualCluster(flexible_resource_min={"CPU": 1}, flexible_resource_max={"CPU": 8}))
+
+# The job uses 1 A100 GPU.
+ray.init(virtual_cluster=VirtualCluster(fixed_size_nodes=[FixedSizeNode(resources={"GPU": 1}, parent_node_selector={"accelerator_type": In("A100")})]))
```

Comment:
what happens if there is no such node exactly matches the fixed size node resource? will we round up, shard the node, or fail?

Reply:
a follow up question is what happens if the required_nodes conflicts with flex_resource_max?

Reply:
A physical node can be split into multiple fixed size virtual nodes.

> a follow up question is what happens if the required_nodes conflicts with flex_resource_max?

flex_resource_max doesn't include fixed size node resources. Total virtual cluster resource is flex resource + fixed resource.

Reply:
would flex_resource_max -> additional_resource make sense?